src/compiler-src.md

```diff
@@ -1,137 +1,191 @@
-incremental compilation. (For details on defining queries, check out
-the [query model].)
-
-Regardless of the general setup, the basic operations that the
-compiler must perform are the same. The only thing that changes is
-whether these operations are invoked front-to-back, or on demand.  In
-order to compile a Rust crate, these are the general steps that we
-take:
-
-1. **Parsing input**
-    - this processes the `.rs` files and produces the AST
-      ("abstract syntax tree")
-    - the AST is defined in `src/librustc_ast/ast.rs`. It is intended to match the lexical
-      syntax of the Rust language quite closely.
-2. **Name resolution, macro expansion, and configuration**
-    - once parsing is complete, we process the AST recursively, resolving
-      paths and expanding macros. This same process also processes `#[cfg]`
-      nodes, and hence may strip things out of the AST as well.
-3. **Lowering to HIR**
-    - Once name resolution completes, we convert the AST into the HIR,
-      or "[high-level intermediate representation]". The HIR is defined in
-      `src/librustc_middle/hir/`; that module also includes the [lowering] code.
-    - The HIR is a lightly desugared variant of the AST. It is more processed
-      than the AST and more suitable for the analyses that follow.
-      It is **not** required to match the syntax of the Rust language.
-    - As a simple example, in the **AST**, we preserve the parentheses
-      that the user wrote, so `((1 + 2) + 3)` and `1 + 2 + 3` parse
-      into distinct trees, even though they are equivalent. In the
-      HIR, however, parentheses nodes are removed, and those two
-      expressions are represented in the same way.
-3. **Type-checking and subsequent analyses**
-    - An important step in processing the HIR is to perform type
-      checking. This process assigns types to every HIR expression,
-      for example, and also is responsible for resolving some
-      "type-dependent" paths, such as field accesses (`x.f` – we
-      can't know what field `f` is being accessed until we know the
-      type of `x`) and associated type references (`T::Item` – we
-      can't know what type `Item` is until we know what `T` is).
-    - Type checking creates "side-tables" (`TypeckTables`) that include
-      the types of expressions, the way to resolve methods, and so forth.
-    - After type-checking, we can do other analyses, such as privacy checking.
-4. **Lowering to MIR and post-processing**
-    - Once type-checking is done, we can lower the HIR into MIR ("middle IR"),
-      which is a **very** desugared version of Rust, well suited to borrowck
-      but also to certain high-level optimizations.
-5. **Translation to LLVM and LLVM optimizations**
-    - From MIR, we can produce LLVM IR.
-    - LLVM then runs its various optimizations, which produces a number of
-      `.o` files (one for each "codegen unit").
-6. **Linking**
-    - Finally, those `.o` files are linked together.
```

Comment:
I like this step-by-step summary. Is there something like it elsewhere in the guide? If so we should link there, if not I think we should keep this.

Reply:
Yes, there is this: https://rustc-dev-guide.rust-lang.org/overview.html

I will add a link.